dom.go

```diff
@@ -252,19 +268,59 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.node.Set("innerHTML", h.innerHTML)
 	}
 
-	return h.reconcileChildren(prev, nil)
+	return h.reconcileChildren(prev)
 }
 
 // reconcileChildren reconciles children of the current HTML against a previous
 // render's DOM nodes.
-func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts []Mounter) {
-	// TODO better list element reuse
+func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
+	var keyed bool
 	for i, nextChild := range h.children {
-		// TODO(pdf): Add tests for node equality
-		new := h.node != prev.node
-		if i >= len(prev.children) || new {
-			if nextChildList, ok := nextChild.(List); ok {
-				nextChildList.reconcile(h.node, insertBefore, nil)
+		// Massage concrete type if necessary.
+		switch v := nextChild.(type) {
+		case *HTML:
+			if v == nil {
+				nextChild = nil
+				h.children[i] = nextChild
+			}
+		case List:
+			nextChild = childList{html: &HTML{children: v}}
+			h.children[i] = nextChild
+		}
+
+		// TODO(pdf): Add tests for node equality, keyed children
+		var (
+			new                   = h.node != prev.node
+			nextKey               string
+			stableKey             bool
+			insertBeforeKeyedNode jsObject
+		)
+		if keyer, ok := nextChild.(Keyer); ok {
+			nextKey = keyer.Key()
+			if len(h.keyMap) > 0 && nextKey == "" {
+				panic("vecty: all siblings must have keys when using keyed elements")
```

Comment:
I agree with the panic, but `nextKey == ""` places an unusual constraint that `""` is not a valid key. As mentioned above, I'd prefer we switch to `interface{}` key types, in which case we can use `nil` here I suppose.

Reply:
Zero value needs to be invalid for whatever type is used, yes.